lib/public_key/src/pubkey_cert.erl

```diff
@@ -18,130 +18,140 @@
+         validate_issuer/4,
          validate_names/6,
-	 validate_extensions/4,
-	 normalize_general_name/1, 
+         validate_signature/6,
+         verify_data/1,
+         verify_fun/4,
+         prepare_for_next_cert/2]).
+
+%% Utility functions
+-export([normalize_general_name/1,
          is_self_signed/1,
-	 is_issuer/2, 
+	 is_issuer/2,
          issuer_id/2,
          subject_id/1,
-         distribution_points/1, 
-	 is_fixed_dh_cert/1, 
-         verify_data/1, 
-         verify_fun/4, 
-	 select_extension/2, 
+         distribution_points/1,
+	 is_fixed_dh_cert/1,
+         select_extension/2,
          match_name/3,
-	 extensions_list/1, 
-         cert_auth_key_id/1, 
-         time_str_2_gregorian_sec/1,
-         gen_test_certs/1, 
+	 extensions_list/1,
+         cert_auth_key_id/1,
+         time_str_2_gregorian_sec/1
+        ]).
+
+%% Generate test data
+-export([gen_test_certs/1,
          x509_pkix_sign_types/1,
-         root_cert/2]).
+         root_cert/2,
+         prune_tree/2]).
 
 -define(NULL, 0).
+-define(EMPTY_POLICY_TREE, [[]]).
 
 %%====================================================================
 %% Internal application APIs
 %%====================================================================
 
-%%--------------------------------------------------------------------
--spec verify_data(DER::binary()) ->
-           {DigestType, PlainText, Signature}
-               when DigestType :: md5 | crypto:sha1() | crypto:sha2() | none,
-                    PlainText  :: binary(),
-                    Signature  :: binary().
-%%
-%% Description: Extracts data from DerCert needed to call public_key:verify/4.
-%%--------------------------------------------------------------------	 
-verify_data(DerCert) ->
-    {ok, OtpCert} = pubkey_cert_records:decode_cert(DerCert),
-    extract_verify_data(OtpCert, DerCert).
+%%====================================================================
+%% Path validation
+%%====================================================================
 
 %%--------------------------------------------------------------------
 -spec init_validation_state(#'OTPCertificate'{}, integer(), list()) ->
 				   #path_validation_state{}.
 %%
 %% Description: Creates initial version of path_validation_state for
 %% basic path validation of x509 certificates.
-%%--------------------------------------------------------------------	 
-init_validation_state(#'OTPCertificate'{} = OtpCert, DefaultPathLen, 
+%%--------------------------------------------------------------------
+init_validation_state(#'OTPCertificate'{} = OtpCert, DefaultPathLen,
 		      Options) ->
-    PolicyTree = #policy_tree_node{valid_policy = ?anyPolicy,
-				   qualifier_set = [],
-				   criticality_indicator = false,
-				   expected_policy_set = [?anyPolicy]},
+    PolicyTree = [[[#{valid_policy  => ?anyPolicy,
+                      qualifier_set => [],
+                      criticality_indicator => false,
```

Comment:
on naming: isn't `critical` informative, good enough name?
on function: does it need to be a included in policy tree nodes at all? I can't figure out the need for it ... please enlight me